judge/views/problem.py

```diff
@@ -201,9 +201,69 @@ def get_context_data(self, **kwargs):
         context['meta_description'] = self.object.summary or metadata[0]
         context['og_image'] = self.object.og_image or metadata[1]
+
+        context['can_vote'] = self.object.can_vote(user)
+        # The vote this user has already cast on this problem.
+        if context['can_vote']:
+            try:
+                context['vote'] = ProblemPointsVote.objects.get(voter=user.profile, problem=self.object)
+            except ObjectDoesNotExist:
+                context['vote'] = None
+        else:
+            context['vote'] = None
+
+        all_votes = list(self.object.problem_points_votes.order_by('points').values_list('points', flat=True))
+        context['has_votes'] = len(all_votes) > 0
+
+        # If the user is not currently in contest.
+        if not self.request.in_contest:
+            context['all_votes'] = all_votes
+
+        context['max_possible_vote'] = settings.DMOJ_PROBLEM_MAX_USER_POINTS_VOTE
+        context['min_possible_vote'] = max(settings.DMOJ_PROBLEM_MIN_USER_POINTS_VOTE,
+                                           settings.DMOJ_PROBLEM_MIN_PROBLEM_POINTS)
         return context
 
 
+class DeleteProblemVote(ProblemMixin, SingleObjectMixin, View):
+    def get(self, request, *args, **kwargs):
+        return HttpResponseForbidden(status=405, content_type='text/plain')
+
+    def post(self, request, *args, **kwargs):
+        self.object = self.get_object()
+        if not request.user.is_authenticated:
+            return HttpResponseForbidden(_('Not signed in.'), content_type='text/plain')
+        elif self.object.can_vote(request.user):
+            ProblemPointsVote.objects.filter(voter=request.profile, problem=self.object).delete()
+            return HttpResponse('success', content_type='text/plain')
+        else:
+            return HttpResponseForbidden(_('Not allowed to delete votes on this problem.'), content_type='text/plain')
+
+
+class ProblemVote(ProblemMixin, SingleObjectMixin, View):
+    def get(self, request, *args, **kwargs):
+        return HttpResponseForbidden(status=405, content_type='text/plain')
+
+    def post(self, request, *args, **kwargs):
+        self.object = self.get_object()
+        if not self.object.can_vote(request.user):
+            return HttpResponseForbidden(_('Not allowed to vote on this problem.'), content_type='text/plain')
+
+        form = ProblemPointsVoteForm(request.POST)
+        if form.is_valid():
+            with transaction.atomic():
+                # Delete any pre existing votes.
+                ProblemPointsVote.objects.filter(voter=request.profile, problem=self.object).delete()
+                vote = form.save(commit=False)
+                vote.voter = request.profile
+                vote.problem = self.object
+                vote.note = vote.note.strip()
+                vote.save()
+            return JsonResponse({'points': vote.points, 'note': vote.note})
+        else:
```

Comment:
There's literally a """resolved""" conversation that says to remove this else. Remove it.